Ignore aborted responses in caching

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->
jdm KiChjang Manishearth Follow up on https://github.com/servo/servo/pull/18676 and https://github.com/servo/servo/pull/19274 to ignore aborted responses in caching. 

I also found out the cache shouldn't return any response whose body is still in `ResponseBody::Receiving` mode, because that fails the assertion at https://github.com/servo/servo/blob/master/components/net/fetch/methods.rs#L438(we might want to add a channel as pat of the cached response later on to deal with this case). I only found out now because I needed the response from the server to trickle in so that it could be cached and aborted. 

I copied the `http-cache.py` server from the wpt folder, and added a 'trickle' option, which is necessary to actually have a failing test with a cached but aborted request, it's now passing. 

I also remove one unused import that slippled through previously.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19350)
<!-- Reviewable:end -->

gterzian committed:
ignore aborted response in http caching

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @manishearth (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/net/fetch/methods.rs, components/net/http_cache.rs, components/net_traits/response.rs, components/net_traits/response.rs

jdm commented on tests/wpt/mozilla/tests/mozilla/resources/http-cache-trickle.py:
```diff
@@ -0,0 +1,64 @@
+
+    response.status = (code, phrase)
+
+    if request.GET.first("trickle", None):
```
If the only change necessary was adding a trickle option, we can get that automatically with the existing script: http://wptserve.readthedocs.io/en/latest/pipes.html#trickle

gterzian replied:
thanks for the pointer. 

I tried applying the trickle option when making the XHR call, however it appears that such 'pipes' are only applied to static files, not python scripts. The docs don't mention this, however in the code it really seems that `wrap_pipeline` is only applied to the response of the `FileHandler`(and therefore not to response of the `PythonScriptHandler`), see https://github.com/w3c/web-platform-tests/blob/master/tools/wptserve/wptserve/handlers.py#L150.
I opened a PR to update the docs by the way, in case this is correct(it would have saved me some debugging time to have a note present) https://github.com/w3c/web-platform-tests/pull/8415

gterzian committed:
handle caching of response with a body of ResponseBody::Receiving

gterzian commented:
Ok, an extra commit to handle `ResponseBody::Receiving`. That one is perhaps a bit of a hack, please take a look. 

To allow the cache to construct a response that hasn't finished receiving it's body yet, the cache simply has to work with the "done channel", in a similar way to the fetch worker thread.

The problem I encountered related to aborting a fetch. Consider the following example:
1. Fetch 1 does a network fetch for resource A. 
2. The response from Fetch 1, with a body of `ResponseBody::Receiving`, is cached.
3. While Fetch 1 is still receiving the body, Fetch 2 starts, for the same resource A.

At this point, if we allow the cache to construct a response with a body of `ResponseBody::Receiving`, we also need to set the done channel to `Some`, otherwise `fetch/method` expects a body in either `of `ResponseBody::Done` or `of `ResponseBody::Empty` (a 'sync' response). 

4. Fetch 2 gets a cached response, and awaits the body. 
5. Fetch 1, before having received the full body, is aborted. 
6. Fetch 2 will not receiving a response with an empty `ResponseBody::Done`. This doesn't reflect the content of the resource that Fetch 2 asked for. There is no way at this point to tell Fetch 2 to go back to the network. 

To handle the problem of 6, I added a loop in `http_loader.rs`, that awaits for either `Data::Done` or `Data::Cancelled`. In the case of Done, the full response is returned, this is the equivalent of having a full response in the cache to begin with. In the case of Cancel, the channel is set back to `None`, and the network fetch takes over. 

Please let me know what you think.

Manishearth commented:
I'm not clear why we need more uses of done_chan here?

gterzian commented:
@Manishearth One could say it's pretty much because of the logic in [wait_for_response](https://github.com/gterzian/servo/blob/95f771f004fce0114bf9840df7d189d96079fb90/components/net/fetch/methods.rs#L422)

gterzian commented:
In the context of this PR, the easy solution is to disregard the second commit, that way we have a cache that only deals in responses that have been fully loaded, and it will simply filter out those that  have been aborted. 

I do think the second commit highlights an interesting problem, perhaps even something that, if we get it right, could help clarify the spec.

The spec makes it obvious that "Typically response’s body’s stream is still being enqueued to after returning"(from within the parallel network fetch steps), and how this fits with caching isn't completely clear. In fact caching is mostly referred to as a fully sync process(can the cache return a response whose body stream is still being enqueued to from another fetch thread?), and the question of a fetch thread being aborted makes this even more interesting. What to do with a response from the cache, whose body is still being enqueued from from another fetch, when that other fetch is suddenly being aborted(it's easy to filter out cached responses that have been already aborted when the subsequent fetch asks for them, it's harder to deal with the "async" situation of the cached response still waiting on the body being fetched by another fetch, who then is aborted)?

gterzian commented:
Have you had time to form thoughts on this one? Looking forward to those...

Regarding the 'trickle' inside the python cache script, I've opened a PR to allow applying pipes to python scripts too. https://github.com/w3c/web-platform-tests/pull/8541

In the meantime, could we perhaps try a test run on this one? It's a bit too slow on my machine, and I'm wondering how the current changes fare against the full suite...

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 95f771f004fce0114bf9840df7d189d96079fb90 with merge ada035253667be45167fb5ad36b0d67379ebd65d...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2810)

KiChjang commented:
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/http-cache-xhr.html:
  │ FAIL [expected PASS] The response from an aborted XHR request should not be cached
  │   → assert_equals: expected 2 but got 0
  │ 
  │ client.onprogress</client2.onloadend</<@http://web-platform.test:8000/_mozilla/mozilla/http-cache-xhr.html:45:11
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1489:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1513:20
```

gterzian commented:
Interesting, that one is passing on my machine, both with debug and release builds, and even if I run the entire mozilla suite. I'll have to look into it...

bors-servo commented:
:umbrella: The latest upstream changes (presumably #19538) made this pull request unmergeable. Please resolve the merge conflicts.

gterzian commented:
could we pls try another test run?

KiChjang commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 1472697ca2d2bb85e792e8914b90aaf4d47b85d2 with merge 30c48f41893f65e3957028ea3dc3102345803a1c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7260)

gterzian commented:
Thanks, so that was the last part of the test failing(as opposed to the first part in the previous run), I removed it because in fact it's unnecessary in this context(it tests for a normal cached response, after the actually 'abort' logic has been already tested). I'm not sure why it failed, on my machine it passed consistently, I can imagine some timing issue only arising when the full suite is run...

I removed that part of the test, could we try another run please? If it passes, it's ready for review...

jdm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b2e4b4041eb692c105d4b006bd87c6a1c6bba4b5 with merge 048b3207d8fb1ea5fcdf03572f39b3a9e64e841f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7271)

gterzian commented:
Ok, interesting, that's an unrelated test that crashed, could this be an intermittent failure? I tried locally and it's passing consistently it seems...

`
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/mozilla/iframe_replacement.html", "line": 93951, "action": "test_result", "expected": "OK"}
`

gterzian commented:
Could we get another test run please? I've actually just rebased off master, and at least it made the previously failing unit tests passing. So far, I haven't been able to reproduce locally the failure for the new wpt test I added. 

This has merged in the meantime https://github.com/w3c/web-platform-tests/pull/8541 so I'll look into updating WPT which will allow us to remove the extra python script, and instead re-use the one in the one already present in the wpt suite and apply the "trickle" argument to it.

Picking this one back up after a while...

Manishearth commented:
@bors-servo try retry

bors-servo commented:
:hourglass: Trying commit 993e2f55eda267c758afd8bbc0e850232b4e7ba0 with merge 302324d79453c97a44d4fe04387135493e7aa58e...

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 993e2f55eda267c758afd8bbc0e850232b4e7ba0 with merge cb526661c33cf4437a52f2a15e4658ad797c4061...

bors-servo commented:
:broken_heart: Test failed - [arm32](http://build.servo.org/builders/arm32/builds/10729)

CYBAI commented:
@bors-servo retry

```
command timed out: 1200 seconds without output running ['./mach', 'build', '--rel', '--target=arm-unknown-linux-gnueabihf'], attempting to kill
```

bors-servo commented:
:hourglass: Trying commit 993e2f55eda267c758afd8bbc0e850232b4e7ba0 with merge f1358de971f770fea26f5be4a2c9753b1fa9446e...

jdm commented:
@CYBAI given that all of the other platforms were building fine (http://build.servo.org/grid) and we don't run tests on arm, retrying because the arm build timed out is not a good use of our resources.

CYBAI commented:
@jdm Sorry for my misunderstanding about `retry`. I will check `grid` first next time! and thanks for the explanation :)

bors-servo commented:
:broken_heart: Test failed - [arm64](http://build.servo.org/builders/arm64/builds/10754)

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 993e2f5 has been approved by `jdm`

<!-- @bors-servo r=jdm 993e2f55eda267c758afd8bbc0e850232b4e7ba0 -->

bors-servo commented:
:hourglass: Testing commit 993e2f55eda267c758afd8bbc0e850232b4e7ba0 with merge ce05a292aa3e0c012680a6a338a11558b963affe...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7652)

CYBAI commented:
From `filtered-wpt-errorsummary.log`,

```
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/webgl/conformance-1.0.3/conformance/more/functions/readPixelsBadArgs.html", "line": 190256, "action": "test_result", "expected": "OK"}
```

From `intermittents.log`,

```
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/css/iframe/hide_after_load.html", "line": 21659, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "/_mozilla/mozilla/simple_scroll_to_fragment.html 2d5103c1d79069b6628ffe931d05456e5b9244b2\n/_mozilla/mozilla/simple_scroll_to_fragment_ref.html 7949e80732fa0f4afdbae3b1cad96ccd75bb04f6\nTesting 2d5103c1d79069b6628ffe931d05456e5b9244b2 == 7949e80732fa0f4afdbae3b1cad96ccd75bb04f6", "stack": null, "subtest": null, "test": "/_mozilla/mozilla/simple_scroll_to_fragment.html", "line": 29691, "action": "test_result", "expected": "PASS"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/webgl/conformance-1.0.3/conformance/glsl/bugs/long-expressions-should-not-crash.html", "line": 120198, "action": "test_result", "expected": "TIMEOUT"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/webgl/conformance-1.0.3/conformance/glsl/misc/shader-with-non-reserved-words.html", "line": 166967, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/webgl/conformance-1.0.3/conformance/more/functions/texImage2DHTML.html", "line": 191242, "action": "test_result", "expected": "OK"}
```

Including #14606, #16216, #18887

jdm commented:
@bors-servo retry
* #19846

bors-servo commented:
:hourglass: Testing commit 993e2f55eda267c758afd8bbc0e850232b4e7ba0 with merge 6d1c9580de28c35d8a8cbd12d2e39654f68bea87...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7655)

jdm commented:
@bors-servo retry
* #19033

bors-servo commented:
:hourglass: Testing commit 993e2f55eda267c758afd8bbc0e850232b4e7ba0 with merge 4307b6e67b0cb35e2afc46ba0b64e7bc5bde1bdf...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10496), [arm32](http://build.servo.org/builders/arm32/builds/10738), [arm64](http://build.servo.org/builders/arm64/builds/10760), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10791), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7657), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7673), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10869), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3269), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3290), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7535), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7603), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3363), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3310), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6173)
Approved by: jdm
Pushing 4307b6e67b0cb35e2afc46ba0b64e7bc5bde1bdf to master...

gterzian commented:
thanks! I'll update the wpt code and remove the then unnecessary `tests/mozilla/resources/http-cache-trickle.py` script in a separate PR...
